cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java

```diff
@@ -50,14 +54,27 @@
  */
 public class DefaultStreamWriter implements StreamWriter {
 
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultStreamWriter.class);
+
   private final String namespaceId;
   private final EndpointStrategy endpointStrategy;
+  /**
+   * The program that is using this {@link StreamWriter}.
+   */
+  private final Id.Program program;
+  private final Map<Id.Stream, Boolean> isStreamRegistered;
+  private final UsageRegistry usageRegistry;
 
   @Inject
   public DefaultStreamWriter(@Assisted("namespaceId") String namespaceId,
+                             @Assisted("programId") Id.Program program,
+                             UsageRegistry usageRegistry,
                              DiscoveryServiceClient discoveryServiceClient) {
     this.namespaceId = namespaceId;
+    this.program = program;
     this.endpointStrategy = new RandomEndpointStrategy(discoveryServiceClient.discover(Constants.Service.STREAMS));
+    this.isStreamRegistered = Maps.newHashMap();
```

Comment:
`write` can be called concurrently, right? If so we should make the map `ConcurrentMap`.